cmd/crio/main.go

```diff
@@ -214,12 +219,24 @@ func main() {
 			logrus.Fatalf("Failed to chmod listen socket %s: %v", config.Listen, err)
 		}
 
+		enableOtel := config.EnableMetrics && config.EnableTracing
+		tracerProvider, otelUnaryInterceptor, otelStreamInterceptor, err := otel.InitOtelTracing(
+			ctx,
+			enableOtel,
+			config.CollectorGRPCPort,
+			config.OtelServiceName,
+			config.TracingExporter,
+		)
 		grpcServer := grpc.NewServer(
 			grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(
 				metrics.UnaryInterceptor(),
 				log.UnaryInterceptor(),
+				otelUnaryInterceptor,
+			)),
+			grpc.StreamInterceptor(grpc_middleware.ChainStreamServer(
+				log.StreamInterceptor(),
+				otelStreamInterceptor,
 			)),
```

Comment:
If `enableOtel == false` then we get `nil` for `otelUnaryInterceptor` and `otelStreamInterceptor`. Registering those values as interceptor will cause CRI-O to panic if the first RPC happens:
```
DEBU[2021-07-22 10:22:20.552858790+02:00] Request: &ListContainersRequest{Filter:&ContainerFilter{Id:,State:&ContainerStateValue{State:CONTAINER_RUNNING,},PodSandboxId:,LabelSelector:map[string]string{},},}  file="go-grpc-middleware/chain.go:25" id=8ac3af45-3109-4ee4-9b51-27a4154c5928 name=/runtime.v1alpha2.RuntimeService/ListContainers
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1d6144d]

goroutine 134 [running]:
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x2542e80, 0xc000967500, 0x210aea0, 0xc00057cb50, 0xc00098b950, 0x1, 0x1, 0x6)
        vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x2d
github.com/cri-o/cri-o/internal/log.UnaryInterceptor.func1(0x2542e80, 0xc000966570, 0x210aea0, 0xc00057cb50, 0xc00072a000, 0xc00072a020, 0x4e30c6, 0x60f92abc, 0x20f3a2b9, 0x4ecb512caaf)
        internal/log/interceptors.go:56 +0x116
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x2542e80, 0xc000966570, 0x210aea0, 0xc00057cb50, 0xc00072a0e0, 0x20, 0x20, 0x204ad40)
        vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63
github.com/cri-o/cri-o/server/metrics.UnaryInterceptor.func1(0x2542e80, 0xc000966570, 0x210aea0, 0xc00057cb50, 0xc00072a000, 0xc00072a040, 0x1d6151a, 0x204ad40, 0xc00072a0e0, 0xc00072a000)
        server/metrics/interceptors.go:24 +0xc6
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x2542e80, 0xc000966570, 0x210aea0, 0xc00057cb50, 0xc000098800, 0x0, 0xc00098bb30, 0x41bb98)
        vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x2542e80, 0xc000966570, 0x210aea0, 0xc00057cb50, 0xc00072a000, 0xc00057ecd8, 0xc000984ba0, 0x52c6a6, 0x20937c0, 0xc000966570)
        vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34 +0xd7
k8s.io/cri-api/pkg/apis/runtime/v1alpha2._RuntimeService_ListContainers_Handler(0x219ed20, 0xc000134df0, 0x2542e80, 0xc000966570, 0xc000136360, 0xc00053e390, 0x2542e80, 0xc000966570, 0xc00055d7b8, 0x6)
        vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.pb.go:8058 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00041ee00, 0x255c4d8, 0xc000550780, 0xc000674000, 0xc00027adb0, 0x332ebb0, 0x0, 0x0, 0x0)
        vendor/google.golang.org/grpc/server.go:1292 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc00041ee00, 0x255c4d8, 0xc000550780, 0xc000674000, 0x0)
        vendor/google.golang.org/grpc/server.go:1617 +0xd05
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000646060, 0xc00041ee00, 0x255c4d8, 0xc000550780, 0xc000674000)
        vendor/google.golang.org/grpc/server.go:940 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        vendor/google.golang.org/grpc/server.go:938 +0x1fd
```

Reply:
right, fixing

Reply:
@saschagrunert I was trying to reproduce the error. Wondering if you set `enableOtel == false` explicitly? We've set the defaults to `true` in `pkg/config/config.go` 
```
		OtelConfig: OtelConfig{
			CollectorGRPCPort: "4317",
			EnableTracing:     true,
			TracingExporter:   "otlp",
			OtelServiceName:   "",
		},
```

Reply:
I think `config.EnableMetrics` is `false` by default, right?